### PR TITLE
Fix: Nested Dicts Handling of Document Metadata

### PR DESCRIPTION
**Pull request by lorenzofavaro**
## Description
When the `MultiQueryRetriever` is used to get the list of documents relevant according to a query, inside a vector store, and at least one of these contain metadata with nested dictionaries, a `TypeError: unhashable type: 'dict'` exception is thrown.
This is caused by the `unique_union` function which, to guarantee the uniqueness of the returned documents, tries, unsuccessfully, to hash the nested dictionaries and use them as a part of key.
```python
unique_documents_dict = {
    (doc.page_content, tuple(sorted(doc.metadata.items()))): doc
    for doc in documents
}
```

## Issue
#9872 (MultiQueryRetriever (get_relevant_documents) raises TypeError: unhashable type: 'dict' with dic metadata)

## Solution
A possible solution is to dump the metadata dict to a string and use it as a part of hashed key.
```python
unique_documents_dict = {
    (doc.page_content, json.dumps(doc.metadata, sort_keys=True)): doc
    for doc in documents
}
```

### CODE REVIEW

File: libs/langchain/langchain/retrievers/multi_query.py [hunk 1]
```diff
@@ -148,7 +149,7 @@ def unique_union(self, documents: List[Document]) -> List[Document]:
         # Create a dictionary with page_content as keys to remove duplicates
         # TODO: Add Document ID property (e.g., UUID)
         unique_documents_dict = {
-            (doc.page_content, tuple(sorted(doc.metadata.items()))): doc
+            (doc.page_content, json.dumps(doc.metadata, sort_keys=True)): doc
```
Comment (baskaryan) [inline libs/langchain/langchain/retrievers/multi_query.py:152]: should we make this configurable? could allow users to pass in a `metadata_encoder` which is used here (and that can have a default value of json.dumps)?  am just trying to get ahead of the case where someone wants to use with non-json-serializable metadata. thoughts @hwchase17 ?

Comment (hwchase17) [reply libs/langchain/langchain/retrievers/multi_query.py:152]: yup nice idea

Comment (lorenzofavaro) [reply libs/langchain/langchain/retrievers/multi_query.py]: Great point, I'm working with PGVector & Chroma and right now they don't support documents with non-json-serializable metadata so I didn't mind. 

Nice to get ahead with it.

Comment (baskaryan) [reply libs/langchain/langchain/retrievers/multi_query.py]: on second thought do we need encoding at all? could we do #10107

**Review by hwchase17 — APPROVED** [review]
probably fine

**Pull request merged**